pubsub/pubsub.go

```diff
@@ -314,15 +364,24 @@ func (s *Subscription) getNextBatch(ctx context.Context) ([]*Message, error) {
 	var q []*Message
 	for _, m := range msgs {
 		id := m.AckID
-		q = append(q, &Message{
+		m2 := &Message{
 			Body:     m.Body,
 			Metadata: m.Metadata,
-			ack: func() {
-				// Ignore the error channel. Errors are dealt with
-				// in the ackBatcher handler.
-				_ = s.ackBatcher.AddNoWait(id)
-			},
-		})
+		}
+		m2.ack = func() {
+			t := float64(time.Since(m2.processingStartTime).Seconds())
+			// Note: m2's mutex is locked here as well. Deadlock will
+			// result if Message.Ack is ever called with s.mu held. That currently
+			// cannot happen, but we should be careful if/when implementing features
+			// like auto-ack.
+			s.mu.Lock()
+			s.avgProcessTime = s.avgProcessTime*(1-decay) + t*decay
```

Comment:
Won't this understimate at startup? E.g., if the first message takes 100ms to start, `avgProcessTime` is 0, and you'll set it to `0* (1-0.05) + .1 * 0.05 = .005` instead of `.1`.

Maybe if `s.avgProcessTime == 0` it should just take `t`?

Reply:
Done.